lib/msal-node/src/client/ManagedIdentityApplication.ts

```diff
@@ -0,0 +1,275 @@
+import { AzureArc } from "./ManagedIdentitySources/AzureArc";
+import { Imds } from "./ManagedIdentitySources/Imds";
+import {
+    ManagedIdentityConfiguration,
+    ManagedIdentityNodeConfiguration,
+    buildManagedIdentityConfiguration,
+} from "../config/Configuration";
+import { version, name } from "../packageMetadata.js";
+import { ManagedIdentityRequest } from "../request/ManagedIdentityRequest";
+import { ManagedIdentityId } from "../config/ManagedIdentityId";
+import { CryptoProvider } from "../crypto/CryptoProvider";
+
+/**
+ * Class to initialize a managed identity and identify the service
+ */
+export class ManagedIdentityApplication {
+    private managedIdentityId: ManagedIdentityId;
+
+    private config: ManagedIdentityNodeConfiguration;
+
+    private logger: Logger;
+    private cacheManager: CacheManager;
+    private networkClient: INetworkModule;
+    private networkManager: NetworkManager;
+    private cryptoProvider: CryptoProvider;
+
+    private identitySource:
+        | ServiceFabric
+        | AppService
+        | CloudShell
+        | AzureArc
+        | Imds;
+
+    private fakeAuthority: Authority;
+
+    constructor(
+        managedIdentityId: ManagedIdentityId,
+        configuration?: ManagedIdentityConfiguration
+    ) {
+        this.config = buildManagedIdentityConfiguration(configuration || {});
+        this.managedIdentityId = managedIdentityId;
+
+        this.logger = new Logger(
+            this.config.system.loggerOptions,
+            name,
+            version
+        );
+
+        this.cacheManager = new DefaultStorageClass(
+            this.managedIdentityId.id,
+            DEFAULT_CRYPTO_IMPLEMENTATION,
+            this.logger
+        );
+
+        this.networkClient = this.config.system.networkClient;
+        this.networkManager = new NetworkManager(
+            this.networkClient,
+            this.cacheManager
+        );
+
+        this.cryptoProvider = new CryptoProvider();
+
+        this.identitySource = this.selectManagedIdentitySource();
+
+        const fakeAuthorityOptions: AuthorityOptions = {
+            protocolMode: ProtocolMode.AAD,
+            knownAuthorities: [Constants.DEFAULT_AUTHORITY],
+            cloudDiscoveryMetadata: "",
+            authorityMetadata: "",
+        };
+        this.fakeAuthority = new Authority(
+            Constants.DEFAULT_AUTHORITY,
+            this.networkClient,
+            this.cacheManager,
+            fakeAuthorityOptions,
+            this.logger
+        );
+    }
+
+    /**
+     * Acquire an access token from the cache or the managed identity
+     * @param managedIdentityRequest
+     * @param managedIdentityId
+     * @returns the access token
+     */
+    public async acquireToken(
+        managedIdentityRequest: ManagedIdentityRequest,
+        managedIdentityId: ManagedIdentityId
+    ): Promise<AuthenticationResult | null> {
+        return (
+            (await this.getCachedToken(
```

Comment:
Think how you can refactor the code from ClientCredentialClient so as to reuse the fetching from cache logic. We should not rewrite this because it is  common between client_credentials flow and MSI. For example, you already have a bug in your code because you don't skip the cache when forceRefresh=true. 

I see 2 ways: 

- create a ClietnCredentialsRequest and set an AppTokenProvider. The AppTokenProvider makes MSI calls etc. Caching is ensured by the ClietnCredentialsRequest.
- OR share code from ClietnCredentialsRequest 

I'd go for the first one, but @neha-bhargava can also suggest.

Reply:
I don't understand this comment. Refactoring the code from ClientCredentialClient is exactly how I wrote the caching logic here. I do skip the cache when forceRefresh is true... on line 144.

Reply:
That is essentially what I've already done, but I left those parts out. I will add them.

Reply:
I think we should reuse the code for getting token from cache and saving the token to cache from client credential. And add conditions to skip the parts not relevant to managed identity.